### PR TITLE
DTS: Convert SAM UART/USART to DT_INST macros (plus pinctrl support)

### PR DESCRIPTION
**Pull request by galak**
Convert Atmel SAM UART & USART to the new DT_INST macros and as part of this conversion we introduce basic pinctrl support for the ATMEL SAM.

### CODE REVIEW

**Comment by nandojve** [comment]
All Atmel drivers need use nodelabel. I'll test this PR on top of #24110.

**Comment by galak** [comment]
> All Atmel drivers need use nodelabel. I'll test this PR on top of #24110.

Why?

**Comment by nandojve** [comment]
> > All Atmel drivers need use nodelabel. I'll test this PR on top of #24110.
> 
> Why?
```
static const struct uart_sam_dev_cfg uart##n##_sam_config = {		\
	.regs = (Uart *)DT_REG_ADDR(DT_NODELABEL(uart##n)),		\
	.periph_id = UART_PROP(n, peripheral_id),			\
	.pin_rx = DT_ATMEL_SAM_PIN(DT_NODELABEL(uart##n), 0),		\
	.pin_tx = DT_ATMEL_SAM_PIN(DT_NODELABEL(uart##n), 1),		\
	UART_SAM_IRQ_CALLBACK(n)					\
};									\
```
The above was tested and it works.

This is an old discussion, people questioned about it on the devicetree slack channel at 2020-04-17, inclusive. Atmel drivers don't work using DT_INST without a property to handle the HW instance. It can't instantiate the proper HW instance using the node instance itself, it simple don't match with HW. If you can't follow, try create a non regular case like disabling USART0 and then try enable only USART1 on SAM_V71_XULT, for instance, then you will see the problem. When you disable USART0 the USART1 will be DT_INST___0 and HW will be wrongly configured.

I already sent a patch that can address this but was declined in favor to use DT_NODELABEL without a technical answer at #23335. The #23338, #23332, #23359 are examples how to use HW instance + DT_INST based on #23335 but were closed in favor to use DT_NODELABEL.

I'm adding people that questioned on slack to see this case. I think is important everyone have this pretty clear.

@pabigot @mnkp @mbolivar 

That said, I've been using since than what seems to be the right solution and sent the #24110 as first PR to collect feedback to continue Atmel drivers conversion.

**Comment by stephanosio** [comment]
@nandojve This PR adds `pinctrl` support and uses that to retrieve pin numbers from the device tree of each U(S)ART instance, so no need to use node label anymore.

**Review by stephanosio — APPROVED** [review]
@galak Thanks for doing this. Now we finally have a proper I/O handling framework in the device tree for SAM series SoC.

**Comment by nandojve** [comment]
> @nandojve This PR adds `pinctrl` support and uses that to retrieve pin numbers from the device tree of each U(S)ART instance, so no need to use node label anymore.
> 
> [#24110 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/24110#pullrequestreview-387896749)

Ohh boy! I was stuck on that and didn't realize!
I tested again and it worked with sam_v71_xult.

File: dts/bindings/serial/atmel,sam-uart.yaml [hunk 1]
```diff
@@ -15,3 +15,16 @@ properties:
...
+
+    pinctrl-0:
+      type: phandles
+      description: |
+        PIO pin configuration for URXD & UTRD signals.  We expect that
+        the phandles will reference pinctrl nodes.  These nodes will
+        have a nodelabel that matches the Atmel SoC HAL defines and
+        be of the form p<port><pin><periph>_<inst>_<signal>.
+
+        For example the USART0 on SAME7x would be
```
Comment (pabigot) [inline dts/bindings/serial/atmel,sam-uart.yaml:27]: I'm not sure how this is wrong, but you're referencing a USART in the atmel,sam-uart YAML, and a UART in the atmel,sam-usart YAML.   In the corresponding devicetree files both devices use the uart compatible.

**Review by pabigot — COMMENTED** [review]
After staring at this long enough I think it makes sense and looks clean.  One anomaly in the documentation for the yaml (seems to cross compatibles).

File: soc/arm/atmel_sam/common/soc_gpio.h [hunk 2]
```diff
@@ -177,4 +177,36 @@ static inline void soc_gpio_debounce_length_set(const struct soc_gpio_pin *pin,
...
+
+/* Construct a soc_gpio_pin element for pin cfg */
+#define ATMEL_SAM_INST_PIN(inst, idx)				\
+	{							\
+		1 << ATMEL_SAM_PIN(inst, idx),			\
+		(Pio *)ATMEL_SAM_PIN_TO_PIO_REG_ADDR(inst, idx),\
+		ATMEL_SAM_PIN_2_PIO_PERIPH_ID(inst, idx),	\
+		ATMEL_SAM_PIN_PERIPH(inst, idx) << 16		\
+	}
+
```
Comment (nandojve) [inline soc/arm/atmel_sam/common/soc_gpio.h:211]: I think this should be an a separated file. This way we will keep all devicetree macros only in one place. In future probably we could add more macros to handle, for instance, clocks.

File: soc/arm/atmel_sam/common/soc_gpio.h [hunk 3]
```diff
@@ -177,4 +177,36 @@ static inline void soc_gpio_debounce_length_set(const struct soc_gpio_pin *pin,
...
+#define ATMEL_SAM_PIN_2_PIO_PERIPH_ID(inst, i) \
+	DT_PROP_BY_PHANDLE(NODE_ID_FROM_PINCTRL_0(inst, i),\
+			   atmel_pins, peripheral_id)
+
+/* Get peripheral cfg associated wiith pinctrl-0 pin at index 'i' */
+#define ATMEL_SAM_PIN_PERIPH(inst, i) \
+	DT_PHA(NODE_ID_FROM_PINCTRL_0(inst, i), atmel_pins, peripheral)
+
+/* Construct a soc_gpio_pin element for pin cfg */
+#define ATMEL_SAM_INST_PIN(inst, idx)				\
```
Comment (nandojve) [inline soc/arm/atmel_sam/common/soc_gpio.h:204]: Why this macros don't have DT_ prefix?

**Review by nandojve — CHANGES_REQUESTED** [review]

File: dts/bindings/serial/atmel,sam-uart.yaml (hunk 1 above)
Comment (galak) [reply dts/bindings/serial/atmel,sam-uart.yaml:27]: :), will fix that up.  copy/paste lots of iterations :)

File: dts/arm/atmel/pinctrl_atmel_sam.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef PINCTRL_ATMEL_SAM_H_
+#define PINCTRL_ATMEL_SAM_H_
+
+#define PERIPH_a	0
+#define PERIPH_b	1
+#define PERIPH_c	2
+#define PERIPH_d	3
```
[the remark this answers is not in the document]
Comment (galak) [reply dts/arm/atmel/pinctrl_atmel_sam.h:12]: Will do.

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 2 above)
Comment (galak) [reply soc/arm/atmel_sam/common/soc_gpio.h:211]: What would you suggest I call the file?

File: dts/arm/atmel/pinctrl_atmel_sam.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (galak) [reply dts/arm/atmel/pinctrl_atmel_sam.h:12]: Done

File: dts/arm/atmel/sam3x-pinctrl.dtsi [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+			DT_ATMEL_PIN(usart1, cts1, a, 15, a);
+			DT_ATMEL_PIN(usart1, rts1, a, 14, a);
+			DT_ATMEL_PIN(usart1, rxd1, a, 12, a);
+			DT_ATMEL_PIN(usart1, sck1, a, 16, a);
+			DT_ATMEL_PIN(usart1, txd1, a, 13, a);
+			DT_ATMEL_PIN(usart2, cts2, b, 23, a);
+			DT_ATMEL_PIN(usart2, rts2, b, 22, a);
+			DT_ATMEL_PIN(usart2, rxd2, b, 21, a);
+			DT_ATMEL_PIN(usart2, sck2, b, 24, a);
+			DT_ATMEL_PIN(usart2, txd2, b, 20, a);
```
[the remark this answers is not in the document]
Comment (galak) [reply dts/arm/atmel/sam3x-pinctrl.dtsi:29]: So the board dtsi at the end is the one that will pick the specific combo's.  The only cases so far that I've put the pinctrl selection in the soc.dtsi is when there is only a single choice for a peripheral.

I think you might be mixing the `pinctrl-0` node with the selection of configuration.   The configuration choices are at the SoC level and effective match the tables in the user manuals.  For example of SAME70, if you look at table 45-2 in section 45.5.1, that is what we have here.

The board will set the `pinctrl-0` in its dts to select which entry in that table makes sense (again only exception is if there is only one possible choice).

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 3 above)
Comment (galak) [reply soc/arm/atmel_sam/common/soc_gpio.h:204]: I want to keep DT_ prefix for things either generator or from devicetree.h headers.  We can call it `ATMEL_SAM_DT_INST_PIN` if you prefer.

File: dts/bindings/serial/atmel,sam-uart.yaml (hunk 1 above)
Comment (galak) [reply dts/bindings/serial/atmel,sam-uart.yaml]: Fixed.

File: dts/arm/atmel/sam3x-pinctrl.dtsi (hunk 5 above)
[the remark this answers is not in the document]
Comment (nandojve) [reply dts/arm/atmel/sam3x-pinctrl.dtsi:29]: Ok lets move this way.

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 2 above)
Comment (nandojve) [reply soc/arm/atmel_sam/common/soc_gpio.h:211]: ../atmel_sam/common/atmel_sam_dt.h

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 3 above)
Comment (nandojve) [reply soc/arm/atmel_sam/common/soc_gpio.h:204]: Ok, It could be ATMEL_SAM_DT_PIN since INST became redundant here, right?

Comment (galak) [reply soc/arm/atmel_sam/common/soc_gpio.h:204]: Will rename to ATMEL_SAM_DT_PIN

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 2 above)
Comment (galak) [reply soc/arm/atmel_sam/common/soc_gpio.h]: Done.

File: soc/arm/atmel_sam/common/soc_gpio.h (hunk 3 above)
Comment (galak) [reply soc/arm/atmel_sam/common/soc_gpio.h]: Done

File: dts/bindings/serial/atmel,sam-usart.yaml [hunk 6]
```diff
@@ -15,3 +15,16 @@ properties:
       type: int
       description: peripheral ID
       required: true
+
+    pinctrl-0:
+      type: phandles
+      description: |
+        PIO pin configuration for URXD & UTRD signals.  We expect that
```
Comment (nandojve) [inline dts/bindings/serial/atmel,sam-usart.yaml:22]: Pins names on USART are rxd and txd.

**Review by nandojve — CHANGES_REQUESTED** [review]

**Review by nandojve — APPROVED** [review]
I don't see anymore and removed the DNM.

File: dts/bindings/serial/atmel,sam-usart.yaml (hunk 6 above)
Comment (galak) [reply dts/bindings/serial/atmel,sam-usart.yaml]: Fixed.

**Pull request merged**